File: tests/python/test_recalled_window_bucket.py
from datetime import datetime, timezone

from starling.bus.bus_event import compute_window_bucket


def _t(epoch: int) -> datetime:
    return datetime.fromtimestamp(epoch, tz=timezone.utc)


def test_buckets_by_2_seconds():
    t0 = _t(1_000_000_000)
    t1 = _t(1_000_000_001)
    t2 = _t(1_000_000_002)
    assert compute_window_bucket("statement.recalled", t0) == \
           compute_window_bucket("statement.recalled", t1)
    assert compute_window_bucket("statement.recalled", t0) != \
           compute_window_bucket("statement.recalled", t2)


def test_expected_string_value():
    assert compute_window_bucket("statement.recalled", _t(1_000_000_000)) == "500000000"

File: tests/cpp/test_recalled_window_bucket.cpp
#include <gtest/gtest.h>
#include "starling/bus/bus_event.hpp"
#include <chrono>

using starling::bus::compute_window_bucket;

TEST(RecalledWindowBucket, BucketsBy2Seconds) {
    using namespace std::chrono;
    auto t0 = system_clock::from_time_t(1'000'000'000);
    auto t1 = system_clock::from_time_t(1'000'000'001);
    auto t2 = system_clock::from_time_t(1'000'000'002);

    EXPECT_EQ(compute_window_bucket("statement.recalled", t0),
              compute_window_bucket("statement.recalled", t1));
    EXPECT_NE(compute_window_bucket("statement.recalled", t0),
              compute_window_bucket("statement.recalled", t2));
}

TEST(RecalledWindowBucket, ExpectedStringValue) {
    using namespace std::chrono;
    auto t = system_clock::from_time_t(1'000'000'000);
    EXPECT_EQ(compute_window_bucket("statement.recalled", t), "500000000");
}
